### ISSUE TITLE
Use keras callbacks for early stopping etc.

### ISSUE THREAD

**Issue by BeWe11** [ludwig-ai/ludwig#524]
I've recently encountered the problem that I would have needed some kind of tolerance value for the early stopping logic, i.e. a minimum improvement value for an improvement to be considered by the early stopping check.

Keras provides an [early stopping callback](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/EarlyStopping) that has several useful parameters (like the mentioned tolerance value). Other callbacks like a [learning rate scheduler](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/LearningRateScheduler) and a [plateau learning rate reducer](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/callbacks/ReduceLROnPlateau) are included as well. These callbacks provide more functionality than their counterparts in ludwig.

I think it might be a good idea to offload these kind of things to keras where possible. Otherwise ludwig will either lack features that ludwig-free tensorflow already has, or time has to be wasted to reimplement existing functionality.

**Comment by w4nderlust** [comment]
Ludwig does not use Keras as a backend, it has a custom way to deal with TensorFlow placeholders and nodes that makes it really difficult to use Keras.
Anyway, we'll try to add the additional functionalities provided by by the callbacks in the near future, thank you for the request.

**Comment by justinxzhao** [comment]
Ludwig has migrated to PyTorch. Early stopping, learning rate scheduling, plateau learning rate reducing, are all implemented in Ludwig 0.5
